controllers/awsmachine_controller.go

```diff
@@ -186,6 +214,20 @@ func (r *AWSMachineReconciler) SetupWithManager(mgr ctrl.Manager, options contro
 			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.AWSClusterToAWSMachines)},
 		).
 		WithEventFilter(pausedPredicates(r.Log)).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					if e.ObjectOld.GetObjectKind().GroupVersionKind().Kind != "AWSMachine" {
+						return true
+					}
+
+					oldMachine := e.ObjectOld.(*infrav1.AWSMachine)
+					newMachine := e.ObjectNew.(*infrav1.AWSMachine)
+
+					return !reflect.DeepEqual(oldMachine.Spec, newMachine.Spec)
+				},
+			},
+		).
```

Comment:
I'm trying to think if there are any cases where we'd want to do an Update when the specs do not match...  The only case I can currently think of is to force a reconciliation when the resource is no longer paused through use of the annotation.

I don't think this would block requeues that are initiated through full resyncs, since I believe those would fall under `Create` rather than `Update`.

I'm not sure if there are other legitimate times where we would want to reconcile on Update regardless of the lack of differences in spec...

Reply:
> I don't think this would block requeues that are initiated through full resyncs since I believe those would fall under Create rather than Update.

Where I should look to double-check this?

Reply:
> Where I should look to double-check this?

I'm not sure of an automated way to check it, but you could spin up the controller overriding the default sync period and using `-v 4` or greater and you could verify through the predicate logging

Reply:
Unless we have a driving reason to add this predicate checking, I think we should likely avoid it for now. At least until we better understand potential implications of it.

Reply:
The reason we added it is to avoid a build up of events without any spec changes that would need to be processed by the reconciler. For AWSMachine, it's just one instance where we manually patch the object. For AWSCluster, there's a few more

Reply:
> The reason we added it is to avoid a build up of events without any spec changes that would need to be processed by the reconciler. For AWSMachine, it's just one instance where we manually patch the object. For AWSCluster, there's a few more

I think my worry is that we are potentially blocking actions where we would want to reconcile.

For example, if the paused annotation is added and then removed from the AWSMachine. With this predicate we would block processing of the AWSMachine when the annotation is removed (the resource is unpaused).

We would also block reconciliation if other metadata changes on the resource (which may be okay).

I mainly want to make sure that we aren't introducing any other potential behavioral changes, since we rely on reconciliation not just for trying to create/update resources based on the spec, but also for exposing the state of the backing AWS resources.

Reply:
We could also check for changes on annotations similar to spec but I understand the broader point of this change potentially introducing unwanted behaviour changes. 

If the predicate function is removed altogether, it may exacerbate any reentrancy bugs we may have. AWSController changes can potentially patch the object 3 times in a single reconcile loop. I wonder if that has the chance to grow exponentially with each of the reconciliation loop patching the objects again.

Reply:
Added a suggestion that should be safer while still avoiding triggering additional reconciliations for incremental status updates

Reply:
I like it 👍  Thanks!